source/RoslynAnalyzers/AsyncAnalyzer.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Octopus.RoslynAnalyzers;
+
+using static Descriptors;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class AsyncAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+        VoidMethodsMustNotBeAsync,
+        MethodsReturningTaskMustBeAsync);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(CheckNode, SyntaxKind.MethodDeclaration);
+    }
+
+    void CheckNode(SyntaxNodeAnalysisContext context)
+    {
+        if (context.Node is not MethodDeclarationSyntax methodDec) return;
+
+        var isAsync = methodDec.Modifiers.Any(SyntaxKind.AsyncKeyword);
+        if (isAsync)
+        {
+            VoidMethods_MustNotBeAsync(context, methodDec);
+        }
+
+        if (methodDec.ReturnType is SimpleNameSyntax { Identifier.Text: "Task" or "ValueTask" })
+        {
+            MethodsReturningTask_MustBeAsync(context, methodDec, isAsync);
+        }
+    }
+
+    void VoidMethods_MustNotBeAsync(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec)
+    {
+        if (methodDec.ReturnType is PredefinedTypeSyntax p && p.Keyword.IsKind(SyntaxKind.VoidKeyword))
+        {
+            context.ReportDiagnostic(Diagnostic.Create(VoidMethodsMustNotBeAsync, methodDec.Identifier.GetLocation()));
+        }
+    }
+
+    void MethodsReturningTask_MustBeAsync(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec, bool isAsync)
+    {
+        // don't flag things that are async (that's good!) or abstract methods.
+        if (isAsync || methodDec.Modifiers.Any(SyntaxKind.AbstractKeyword)) return;
+
+        var declaringType = methodDec.Parent;
+        while (declaringType is not null && declaringType is not TypeDeclarationSyntax) declaringType = declaringType.Parent;
+
+        if (declaringType is InterfaceDeclarationSyntax) return; // don't flag on interfaces
+
+        // exemption for classes implementing IAsyncApiAction
```

Comment:
Why do we have this exemption?

Reply:
Because the reflection-based convention test had the exemption and if I disable it I get about 30 violations

Why was THAT exemption there? I don't know and the reflection-based test never mentioned it?

Reply:
I reckon that's a good thread to pull on.
I think I saw elsewhere that suggested that `IAsyncApiAction` might be related to nancy 🤔

Reply:
Yeah it is.

From the looks of the actual code, there doesn't seem to be any reason why those things can't be fixed to not need the exemption, but it'd take a while to go through them all. Let us summon @andrewabest  who is most likely to know about this

Reply:
[Aside: I don't want to hold up merging this for days and days; AsyncApiAction work can follow later]